crates/router/src/utils/user/two_factor_auth.rs

```diff
@@ -202,3 +139,88 @@ pub async fn delete_recovery_code_from_redis(
         .change_context(UserErrors::InternalServerError)
         .map(|_| ())
 }
+
+fn get_totp_attempts_key(user_id: &str) -> String {
+    format!("{}{}", consts::user::REDIS_TOTP_ATTEMPTS_PREFIX, user_id)
+}
+fn get_recovery_code_attempts_key(user_id: &str) -> String {
+    format!(
+        "{}{}",
+        consts::user::REDIS_RECOVERY_CODE_ATTEMPTS_PREFIX,
+        user_id
+    )
+}
+
+pub async fn insert_totp_attempts_in_redis(state: &SessionState, user_id: &str) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+    let user_attempts_value = get_totp_attempts_from_redis(state, user_id).await?;
+    redis_conn
+        .set_key_with_expiry(
+            &get_totp_attempts_key(user_id),
+            user_attempts_value + 1,
+            consts::user::REDIS_TOTP_ATTEMPTS_TTL_IN_SEC,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+pub async fn get_totp_attempts_from_redis(state: &SessionState, user_id: &str) -> UserResult<u8> {
+    let redis_conn = super::get_redis_connection(state)?;
+    redis_conn
+        .get_key::<Option<u8>>(&get_totp_attempts_key(user_id))
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .map(|v| v.unwrap_or(0))
+}
+
+pub async fn insert_recovery_code_attempts_in_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+
+    let user_attempts_value = get_recovery_code_attempts_from_redis(state, user_id).await?;
+    redis_conn
+        .set_key_with_expiry(
+            &get_recovery_code_attempts_key(user_id),
+            user_attempts_value + 1,
+            consts::user::REDIS_RECOVERY_CODE_ATTEMPTS_TTL_IN_SEC,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+
+pub async fn get_recovery_code_attempts_from_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<u8> {
+    let redis_conn = super::get_redis_connection(state)?;
+    redis_conn
+        .get_key::<Option<u8>>(&get_recovery_code_attempts_key(user_id))
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .map(|v| v.unwrap_or(0))
+}
+
+pub async fn delete_totp_attempts_from_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+    redis_conn
+        .delete_key(&get_totp_attempts_key(user_id))
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .map(|_| ())
+}
+
+pub async fn delete_recovery_code_attempts_from_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+    redis_conn
+        .delete_key(&get_recovery_code_attempts_key(user_id))
```

Comment:
Can you check if this breaks if key is not present in redis.

Reply:
Tried for skip , it didn't throw any error